src/libostree/ostree-sysroot-deploy.c

```diff
@@ -2824,34 +2837,250 @@ get_var_dfd (OstreeSysroot      *self,
+    return NULL;
+  return g_file_resolve_relative_path (root, path);
+}
+
 static gboolean
-sysroot_finalize_deployment (OstreeSysroot     *self,
-                             OstreeDeployment  *deployment,
-                             OstreeDeployment  *merge_deployment,
-                             GCancellable      *cancellable,
-                             GError           **error)
+add_kargs_from_filesystem (GTree         *kargs_configs,
+                           int            deployment_dfd,
+                           const char    *dirpath,
+                           GCancellable  *cancellable,
+                           GError       **error)
 {
-  g_autofree char *deployment_path = ostree_sysroot_get_deployment_dirpath (self, deployment);
-  glnx_autofd int deployment_dfd = -1;
-  if (!glnx_opendirat (self->sysroot_fd, deployment_path, TRUE, &deployment_dfd, error))
+  int kargs_dfd = -1;
+  if (!ot_openat_ignore_enoent (deployment_dfd, dirpath,
+      &kargs_dfd, error))
     return FALSE;
+  if (kargs_dfd > 0)
+    {
+      g_auto(GLnxDirFdIterator) dfd_iter = { 0, };
+      if (!glnx_dirfd_iterator_init_at (kargs_dfd, ".", TRUE, &dfd_iter, error))
+        return FALSE;
+      while (TRUE)
+        {
+          struct dirent *dent = NULL;
+          if (!glnx_dirfd_iterator_next_dent_ensure_dtype (&dfd_iter, &dent,
+                                                           cancellable, error))
+            return FALSE;
+          if (dent == NULL)
+            break;
+
+          if (!(dent->d_type == DT_REG || dent->d_type == DT_LNK))
+            continue;
+
+          const char *karg_name = dent->d_name;
+          if (g_tree_lookup (kargs_configs, karg_name))
+            continue;
+
+          g_autofree char *karg_contents = NULL;
+          if (!(karg_contents = glnx_file_get_contents_utf8_at (kargs_dfd,
+                                                                karg_name, NULL,
+                                                                cancellable,
+                                                                error)))
+            return FALSE;
 
+          g_tree_insert (kargs_configs, g_strdup (karg_name),
+                         g_steal_pointer (&karg_contents));
+        }
+    }
+
+  return TRUE;
+}
+
+static gboolean
+add_kargs_from_commit (OstreeSysroot *sysroot,
+                       GTree         *kargs_configs,
+                       int            deployment_dfd,
+                       const char    *dirpath,
+                       const char    *revision,
+                       GCancellable  *cancellable,
+                       GError       **error)
+{
+  struct stat stbuf;
+  if (!glnx_fstatat_allow_noent (deployment_dfd, dirpath, &stbuf, 0, error))
+    return FALSE;
+  const gboolean dir_exists = (errno == 0);
+
+  if (dir_exists)
+    {
+      g_autoptr(GFile) karg_file = get_file_from_repo (ostree_sysroot_repo (sysroot),
```

Comment:
Would it be clearer to name `karg_file` something like `karg_dir` to clearly distinguish it from the children files in `usr/lib/ostree-boot/kargs.d` that contain the kargs?